README.md
# otl_interpreter

Otl interpreter for platform architecture 2.0. Plugin for [complex rest](https://github.com/ISGNeuroTeam/complex_rest/tree/develop)

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

###  Prerequisites
1. [Miniconda](https://docs.conda.io/en/latest/miniconda.html)
2. [Conda-Pack](https://conda.github.io/conda-pack)

### Deploy
1. Clone [complex rest](https://github.com/ISGNeuroTeam/complex_rest/tree/develop)
```bash
git clone git@github.com:ISGNeuroTeam/complex_rest.git
cd ./complex_rest
git checkout develop
```
2. Make `plugin_dev` and `plugins_dir`
```bash
mkdir plugins
mkdir plugin_dev
```
3. Clone otl_interpreter in `plugin_dev`
```bash
cd ./plugin_dev
git clone git@github.com:ISGNeuroTeam/otl_interpreter.git
```
4. Create python virtual environment
```bash
cd ./plugin_dev/otl_interpreter/otl_interpreter/
conda create -p ./venv
conda install -p ./venv python==3.9.7 -y
conda activate ./venv
pip install -r ../requirements.txt
```
5. Create symlink to otl_interpreter in `plugins` directory
```bash
cd ./plugins
ln -r -s ../plugin_dev/otl_interpreter/otl_interpreter otl_interpreter
```
6. Create configuration file
```bash
cd ./plugin_dev/otl_interpreter/otl_interpreter
cp otl_interpreter.conf.example otl_interpreter.conf
```
7. Create complex rest development environment in complex_rest directory
```bash
make dev
```
8. Launch complex_rest
```bash
./start.sh
```
9. Create database for otl_interpreter. Connect to postgres. User: `postgres`. Password: `postgres`
```bash
psql -U postgres -h localhost -p 5433
```
```postgresql
create role otl_interpreter with password 'otl_interpreter';
alter role otl_interpreter with login;
create database otl_interpreter;
grant all privileges on database otl_interpreter to otl_interpreter;
```
10. Make migrations and migrate them to database
```bash
source ./venv/bin/activate
python ./complex_rest/manage.py makemigrations otl_interpreter
python ./complex_rest/manage.py migrate otl_interpreter --database otl_interpreter
```

11. Checkout:  
- http://localhost:8080/admin/. Login: `admin`. Password: `admin`.  
- http://localhost:8080/otl_interpreter/v1/checkjob/


## Running the tests
### Run tests in docker  
```bash
./plugin_dev/otl_interpreter/docs/scripts/run_tests_in_docker.sh
```
### Run tests
1. Launch complex_rest environment
```bash
./start.sh
```
2. Stop dispatcher process
```bash
source ./venv/bin/activate
supervisorctl stop dispatcher
```
3. Run tests
```bash
python ./complex_rest/manage.py test ./plugin_dev/otl_interpreter/tests --settings=core.settings.test
```


## Deployment

* Make plugin archive:
```bash
make pack
```
* Unpack archive into complex_rest plugins directory
* Create database for otl_interpreter and make migrations
* Make otl_interpreter config and configure it
```bash
cp ./otl_interpreter/otl_interpreter.conf.example ./otl_interpreter/otl_interpreter.conf
```
* Make ot_simple_rest_job_proxy config and configure it 
```bash
cp ./ot_simple_rest_job_proxy/ot_simple_rest_job_proxy.conf.example ./ot_simple_rest_job_proxy/ot_simple_rest_job_proxy.conf
```



## Built With

* [Django](https://docs.djangoproject.com/en/3.2.7/) - The web framework used


## Contributing

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/your/project/tags). 

## Authors


## License

[OT.PLATFORM. License agreement.](LICENSE.md)

## Acknowledgments

* Hat tip to anyone whose code was used
* Inspiration
* etc

CHANGELOG.md
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.3.5] - 2023-12-04
### Fixed
- Fixed tasks freezing if there was no message in the exception
### Added
- Added `docker_dev` target in Makefile

## [1.3.4] - 2023-08-03
### Fixed
- Fixed `Otljob not found` error in ot_simple_rest_job_proxy
### Added
- Plugin version logging at startup

## [1.3.3] - 2023-02-16
### Fixed
- Add task that  removes forever calculating results

## [1.3.2] - 2023-02-06
### Fixed
- Fixed integrity error when two equal otl jobs created simultaneously 

## [1.3.1] - 2023-01-31
### Changed
- When node job failed put only node job uuid in status text instead of entire body
- Computing node activated when getting resource info 
- Set default encoding to utf-8 in ot_simple_rest_job_proxy

### Fixed
- Fixed error when nodejob canceled and awated results  nodejobs exists
- Delete otl job info from redis when otl job deleted


## [1.3.0] - 2022-12-02
### Added
- Add user guid in node job dictionary
- Add status filters for node job and otl job in django admin

## [1.2.2] - 2022-11-23
### Fixed
- Fixed error with complex_rest authentication in ot_simple_rest_job_proxy

## [1.2.1] - 2022-11-22
### Changed 
- Download miniconda 4.12 in local directory while building

## [1.2.0] - 2022-11-22
### Changed
- Using otlang version 1.2.1
### Added
- Add set_cache command.
- Add node job uuid in admin list display 
- Add complex rest authentication support to ot_simple_rest_job_proxy
### Fixed
- Fix otl job finished status before executing root job
- Fix ot_simple_rest_job_proxy views. Using ordinary django views.

## [1.1.0] - 2022-09-05
### Changed
- Job proxy checkjob view accepts POST requests
- Adapt otl interpreter to 1.2.0 otlang version
### Added
- Running tests in docker

## [1.0.5] - 2022-06-20
### Fixed
- Fix `__timestamp__` condition
- Fix mistyping in checking result status

## [1.0.4] - 2022-05-30
### Changed
- `use_timewindow` flag stored in NodeCommand model now
### Added
- Add `idempotent` flag to command model. If `False` then `__timestamp__` argument will be added in parsed command.

## [1.0.3] - 2022-05-23
### Fixed
- Planning node job for inactive computing nodes was fixed
- Fixed empty computing node pool when dispatcher restarted
### Added
- Canceling otl job by timeout
- Accepting url encoded data in simple rest proxy plugin
### Changed
- Job planner always creates node job for async subsearch 

## [1.0.2] - 2022-04-27
### Fixed
- Allow node job status transition from READY_TO_EXECUTE to WAITING_SAME_RESULT

## [1.0.1] - 2022-04-26
### Fixed
- OTLang too precise version

## [1.0.0] - 2022-04-25
### Added
- Proxy plugin for ot_simple_rest
- Result dataframe managing 
- Cancel job endpoint
### Fixed
- Mock computing node accepts canceled status
- Fix disappearing nodes from computing node pool when dispatcher restarts
- Change job status to FAILED when syntaz error
### Changed
- Otlang version to 1.1.0

## [0.1.0] - 2022-03-29
### Added
- Changelog.md.

### Changed
- Start using "changelog".


otl_interpreter/settings.py
import configparser
import datetime
import os

from pathlib import Path
from core.settings.ini_config import merge_ini_config_with_defaults


default_ini_config = {
    'logging': {
        'level': 'INFO'
    },
    'db_conf': {
        'host': 'localhost',
        'port': '5433',
        'database':  'otl_interpreter',
        'user': 'otl_interpreter',
        'password': 'otl_interpreter'
    },
    'node_job': {
      'cache_ttl': '60'
    },
    'job_planer': {
        'computing_node_type_priority': 'SPARK EEP POST_PROCESSING',
        'subsearch_is_node_job': 'True'
    },
    'dispatcher': {
        'one_process_mode': 'False',
        'check_job_queue_period': '10',
        'host_id': os.popen("hostid").read().strip(),
        'health_check_period': '15'
    },
    'otl_job_defaults': {
        'cache_ttl': '60',
        'timeout': '0',
        'shared_post_processing': 'True',

    },
    'result_managing': {
        'data_path': 'data',
        'schema_path': '_SCHEMA',
    },
    'service_task_options': {
        'remove_expired_dataframes_period': '15',
        'keep_query_info_days': '30'
    },
    'storages': {
        'shared_post_processing': '/opt/otp/shared_storage',
        'local_post_processing': '/opt/otp/local_storage',
        'interproc_storage': '/opt/otp/inter_proc_storage',
    }
}

PROJECT_DIR = Path(__file__).parent

# try to read path to config from environment
conf_path_env = os.environ.get('otl_interpreter_conf', None)

if conf_path_env is None:
    conf_path = PROJECT_DIR / 'otl_interpreter.conf'
else:
    conf_path = Path(conf_path_env).resolve()

config_parser = configparser.ConfigParser()

config_parser.read(conf_path)

ini_config = merge_ini_config_with_defaults(config_parser, default_ini_config)

job_planer_config = ini_config['job_planer']

host_id = ini_config['dispatcher']['host_id']

DATABASE = {

        "ENGINE": 'django.db.backends.postgresql',
        "NAME": ini_config['db_conf']['database'],
        "USER": ini_config['db_conf']['user'],
        "PASSWORD": ini_config['db_conf']['password'],
        "HOST": ini_config['db_conf']['host'],
        "PORT": ini_config['db_conf']['port']
}

CELERY_BEAT_SCHEDULE = {
    'delete_expired_results': {
        'schedule': datetime.timedelta(
            seconds=int(ini_config['service_task_options']['remove_expired_dataframes_period'])
        ),
        'task': 'otl_interpreter.tasks.delete_expired_results',
    },
    'remove_old_otl_query_info_from_db': {
        'schedule': datetime.timedelta(
            days=int(ini_config['service_task_options']['keep_query_info_days'])
        ),
        'task': 'otl_interpreter.tasks.remove_old_otl_query_info_from_db',
    },
    'remove_forever_calculating_results': {
        'schedule': datetime.timedelta(
            minutes=5
        ),
        'task': 'otl_interpreter.tasks.remove_forever_calculating_results',
    },
    'cancel_otl_job_by_timeout': {
        'schedule': datetime.timedelta(
            seconds=15
        ),
        'task': 'otl_interpreter.tasks.cancel_otl_job_by_timeout',
    }
}


def get_cache():
    # plugin settings load before complex_rest settings. redis is not configured so
    # import must be here
    from cache import get_cache as complex_rest_get_cache
    return complex_rest_get_cache(
        'RedisCache', namespace='otl_interpreter', timeout=300, max_entries=300
    )
